Extremely Tentative El Capitan Support

DomT4 opened the pull request:
Re https://github.com/Homebrew/homebrew/issues/40519#issuecomment-110456475

I *literally* added a caveat that it is unsupported, since we've started getting heavily-commented issues already.

DomT4 commented:
CC @Homebrew/owners @bfontaine

mistydemeo commented on Library/Homebrew/os/mac.rb:
```diff
@@ -214,6 +214,7 @@ def preferred_arch
       "6.3"   => { :clang => "6.1", :clang_build => 602 },
       "6.3.1" => { :clang => "6.1", :clang_build => 602 },
       "6.3.2" => { :clang => "6.1", :clang_build => 602 },
+      "7.0"   => { :clang => "6.1", :clang_build => 602 },
```
I'm seeing `7.0.0 (clang-700.0.53)` with the CLT.

DomT4 replied:
:+1:, Will tweak.

DomT4 replied:
Xcode has it down as `Version 7.0 beta (7A120f)`, but I'm happy to go with the CLT.

mistydemeo replied:
Sorry, I mean the clang version I'm seeing is different from the one you reported - I assume 7.0 beta has the same version?

DomT4 replied:
Unless you mean the `clang_build` should be `700`? Apologies for the confusion :smile_cat:

mistydemeo replied:
Yeah, exactly - the clang I see is labelled 7.0.0 build 700, so assuming it's the same in Xcode, we want `{ :clang => "7.0", :clang_build => 700 }`

DomT4 replied:
Thanks, Sorry for the confusion there. You're right, I was looking at the wrong thing:

```
/Volumes/Xcode/Xcode-beta.app/Contents/Developer/usr/bin/gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 6.1.0 (clang-602.0.53) (based on LLVM 3.6.0svn)
Target: x86_64-apple-darwin14.4.0
Thread model: posix
```

I should have been looking at:

```
/Volumes/Xcode/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang --version
Apple LLVM version 7.0.0 (clang-700.0.53)
Target: x86_64-apple-darwin14.4.0
Thread model: posix
```

I presume?

mistydemeo replied:
Yeah, the latter looks correct. Kind of weird the "gcc" shim points at an older clang though.

abradner commented:
Compiling packages breaks (for me at least) unless you are on the Xcode 7 Beta. While this is mostly obvious, I forgot to `xcode-select` to the new xcode beta and was scratching my head for a while until I spotted my mistake. 
If we're warning users about El Capitan, it might be worthwhile to have `brew doctor` complain if your xcode 7 isn't set up correctly

mikemcquaid commented on Library/Homebrew/os/mac/xcode.rb:
```diff
@@ -15,10 +15,11 @@ def latest_version
         when "10.8"  then "5.1.1"
         when "10.9"  then "6.2"
         when "10.10" then "6.3.2"
+        when "10.11" then "7.0"
```
Might not need this if it's in the `else`

DomT4 replied:
We made this change last year, but I'm happy to drop it?

DomT4 replied:
Killed it.

mikemcquaid commented on Library/Homebrew/cmd/doctor.rb:
```diff
@@ -428,6 +428,14 @@ def check_tmpdir_sticky_bit
+    Homebrew can resolve those issues until El Capitan is stable.
+  EOS
+  end
+end
```
When is this printed? Might be worth printing it on literally every `brew` run.

DomT4 replied:
Every and any time a `brew doctor` is run. I can move it to `brew.rb` instead though? We already have a too-old OS X version warning in there, so a warning about too-new wouldn't be out of place.

mikemcquaid commented:
Would be good to confirm all these values on a VM or actual machine by one of the maintainers.

vszakats commented on Library/Homebrew/cmd/doctor.rb:
```diff
@@ -428,6 +428,14 @@ def check_tmpdir_sticky_bit
 
+def check_supported_osx
+  if MacOS.version == "10.11" then <<-EOS.undent
+    Although Homebrew has added initial recognition of 10.11 it is unsupported.
```
Would there be a comma appropriate here?:
Although Homebrew has added initial recognition of 10.11**,** it is unsupported.

DomT4 replied:
Yup, no problem with that suggestion.

DomT4 commented on Library/brew.rb:
```diff
@@ -50,6 +50,15 @@
+    Although Homebrew has added initial recognition of 10.11, it is unsupported.
+    You may encounter breakage or other failure and there is no guarantee
+    Homebrew can resolve those issues until El Capitan is stable.
+
```
Blank line here to make sure the commands aren't buffered against each other, i.e.

```
~> brew info openssl
Warning: Although Homebrew has added initial recognition of 10.11, it is unsupported.
You may encounter breakage or other failure and there is no guarantee
Homebrew can resolve those issues until El Capitan is stable.

openssl: stable 1.0.2a-1 (bottled)
OpenSSL SSL/TLS cryptography library
https://openssl.org
```

Rather than:

```
~> brew info openssl
Warning: Although Homebrew has added initial recognition of 10.11, it is unsupported.
You may encounter breakage or other failure and there is no guarantee
Homebrew can resolve those issues until El Capitan is stable.
openssl: stable 1.0.2a-1 (bottled)
OpenSSL SSL/TLS cryptography library
https://openssl.org
```

DomT4 commented:
Addressed all feedback in new push.

* Clang version corrected.
* Unsupported warning moved from Doctor to every time `brew` is executed.
* Added a comma to the initial warning sentence.

DomT4 commented:
@Homebrew/owners Any objection to shipping this? We're getting a lot of errors because people haven't updated their Xcode to match OS X 10.11, which this should kick up a `brew doctor` warning about.

jacknagel commented:
I imagine enough people are going to complain that we'll have to move the 10.11 warning to a doctor-only warning, but I guess we can try it.

DomT4 commented:
I moved it to each run per Mike's request, but I don't mind where we put it. Whatever the consensus is.

mistydemeo commented on Library/brew.rb:
```diff
@@ -50,6 +50,20 @@
   EOABORT
 end
 
+if OS.mac? && MacOS.version == "10.11" && MacOS::Xcode.version < "7.0"
```
Need a more complex conditional here - we need to check for the cases where:

* Xcode is installed, but below 7
* Xcode isn't installed, but CLT 7 is installed
* Xcode and CLT are both installed, both out of date
* Xcode and CLT are both installed, only one is up to date

This should probably also be part of the Xcode preinstall checks in `cmd/install` rather than here: https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/cmd/install.rb#L112-L123

DomT4 replied:
I thought @mikemcquaid wanted it runtime for `brew`, rather than just on `install`, but I could have misunderstood :smile_cat:.

Agreed that the conditional is too simplistic in this situation; Should've realised that before I pushed really. Will tune it up to cover the wider scope, thanks :)

jacknagel replied:
I believe the Xcode/CLT doctor checks handle all these cases already. I don't see the benefit of duplicating all of that logic here.

jacknagel replied:
i.e., it doesn't really make sense for things like `brew --prefix` or `brew info` or `brew deps` to outright fail if Xcode is outdated. We already have pre-install checks for this situation.

DomT4 replied:
Maybe the best solution here is to revert it back to what it was before the latest push, merge that, and then take firmer action if the number of issues created because of non-system-appropriate-xcode/clt doesn't drop?

DomT4 replied:
Reverted.

If we want to issue a sterner warning I'm happy to use the existing Doctor methods, but attempt to make them more fatal on 10.11 (i.e. you cannot try and install xyz until your Xcode/CLT is up-to-date), but the consensus seems to lean towards seeing how much the basic support reduces the level of Issues before we go more drastic.

If that's the case, this should be good to :rocket: once it's green on the bot again. I'll let one of you pull the trigger on this if you're happy.

jacknagel commented:
Can we merge at least the first commit, please, so that doctor can start warning 10.11 users that they need to update Xcode/CLT?

DomT4 commented:
I merged it whole. If we get complaints about the runtime warning being too frequent I'll move it back to `doctor`, but there seems to be a basic consensus about getting this in to reduce the level of 10.11 issues solved by updating CLT/Xcode.
